src/map/mob.c

```diff
@@ -1903,12 +1903,41 @@ static int mob_ai_hard(int tid, int64 tick, int id, intptr_t data)
 /*==========================================
  * Initializes the delay drop structure for mob-dropped items.
  *------------------------------------------*/
-static struct item_drop *mob_setdropitem(int nameid, int qty, struct item_data *data)
+static struct item_drop *mob_setdropitem(int nameid, struct mob_drop_option *options, int qty, struct item_data *data)
 {
 	struct item_drop *drop = ers_alloc(item_drop_ers, struct item_drop);
 	drop->item_data.nameid = nameid;
 	drop->item_data.amount = qty;
 	drop->item_data.identify = data ? itemdb->isidentified2(data) : itemdb->isidentified(nameid);
+	
+	// Set item options [KirieZ]
+	if (options != NULL) {
+		int i;
+		
+		for (i = 0; i < options->options_count; i++) {
+			int idx = 0;
+			int min, max;
+			int count;
+
+			if (rnd() % 10000 >= options->options_rate[i])
+				continue;
+
+			// count avoids a too long loop that would cause lag. if after 10 full iterations it still fails
+			// it'll stop in one random index in the next one.
+			count = 10 * options->options[i].option_count + (rnd() % options->options[i].option_count);
```

Comment:
I didn't like this count, I think it kind of breaks the idea of rates, but I couldn't really think of a better way. One idea was to make something like we do with boxes, where we put how many times that item appears in the set.